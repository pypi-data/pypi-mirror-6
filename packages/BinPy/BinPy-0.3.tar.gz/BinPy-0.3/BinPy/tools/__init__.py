from BinPy.tools.digital import *
from BinPy.tools.powersource import *
from BinPy.tools.ground import *
