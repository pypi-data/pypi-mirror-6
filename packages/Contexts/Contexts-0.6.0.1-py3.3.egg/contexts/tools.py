import time as time_module
import sys


def catch(func, *args, **kwargs):
    """
    Call the supplied function with the supplied arguments,
    catching and returning any exception that it throws.

    Arguments:
        func: the function to run.
        *args: positional arguments to pass into the function.
        **kwargs: keyword arguments to pass into the function.
    Returns:
        If the function throws an exception, return the exception.
        If the function does not throw an exception, return None.
    """
    try:
        func(*args, **kwargs)
    except Exception as e:
        return e


def time(func, *args, **kwargs):
    """
    Call the supplied function with the supplied arguments,
    and return the total execution time as a float in seconds.

    The precision of the returned value depends on the precision of
    `time.time()` on your platform.

    Arguments:
        func: the function to run.
        *args: positional arguments to pass into the function.
        **kwargs: keyword arguments to pass into the function.
    Returns:
        Execution time of the function as a float in seconds.
    """
    start_time = time_module.time()
    func(*args, **kwargs)
    end_time = time_module.time()
    return end_time - start_time


def set_trace():
    """Start a Pdb instance at the calling frame, with stdout routed to sys.__stdout__."""
    # https://github.com/nose-devs/nose/blob/master/nose/tools/nontrivial.py
    import pdb
    pdb.Pdb(stdout=sys.__stdout__).set_trace(sys._getframe().f_back)


def setup(func):
    """
    Decorator. Marks a method as a setup method.
    """
    func._contexts_role = "establish"
    return func


def action(func):
    """
    Decorator. Marks a method as an action method.
    """
    func._contexts_role = "because"
    return func


def assertion(func):
    """
    Decorator. Marks a method as an assertion method.
    """
    func._contexts_role = "should"
    return func


def teardown(func):
    """
    Decorator. Marks a method as a teardown method.
    """
    func._contexts_role = "cleanup"
    return func


def spec(cls):
    """
    Class decorator. Marks a class as a test class.
    """
    cls._contexts_role = "Spec"
    return cls
