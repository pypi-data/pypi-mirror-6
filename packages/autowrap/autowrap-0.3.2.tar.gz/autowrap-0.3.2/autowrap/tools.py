

# for backwards compatibility with python 2.6 we use our own implementation
# of an OrderedDictionary, which is not feature complete but simple and
# sufficient for our uses

class OrderKeepingDictionary(object):

    def __init__(self):
        self._dd  = dict()
        self._keys = []

    def __setitem__(self, k, v):
        if k not in self._keys:
            self._keys.append(k)
        self._dd[k] = v

    def __getitem__(self, k):
        return self._dd[k]

    def update(self):
        raise NotImplementedError("update not implemented")

    def iterkeys(self):
        for k in self._keys:
            yield k

    __iter__ = iterkeys

    def itervalues(self):
        for k in self._keys:
            yield self._dd[k]

    def iteritems(self):
        for k in self._keys:
            yield (k, self._dd[k])

    def keys(self):
        return list(self.iterkeys())

    def values(self):
        return list(self.itervalues())

    def items(self):
        return list(self.iteritems())

    def get(self, k, default=None):
        return self._dd.get(k, default)

    def setdefault(self, k, default):
        if k not in self._keys:
            self._keys.append(k)
        return self._dd.setdefault(k, default)

    def __delitem__(self, k):
        raise NotImplementedError("__delitem__ not implemented")

    def __len__(self):
        return len(self._dd)





