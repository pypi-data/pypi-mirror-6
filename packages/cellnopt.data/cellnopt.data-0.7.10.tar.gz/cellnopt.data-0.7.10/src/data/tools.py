# -*- python -*-
#
#  This file is part of the CNO software
#
#  Copyright (c) 2011-2012 - EBI
#
#  File author(s): Thomas Cokelaer <cokelaer@ebi.ac.uk>
#
#  Distributed under the GPLv2 License.
#  See accompanying file LICENSE.txt or copy at
#      http://www.gnu.org/licenses/gpl-2.0.html
#
#  CNO website: www.cellnopt.org
#
##############################################################################
"""Common tools to get the correct filename and pathname. """
from os.path import join as pj
import os
import glob
import urllib

__all__ = ['cnodata', 'checksum']

import logging
logging.basicConfig(level=logging.INFO)


# You can add as much url as desired in baseurls
baseurls = ["http://www.ebi.ac.uk/~cokelaer/cno/doc/sampleModels/_downloads/"]


def cnodata(filename, verbose=True, local=False):
    """Returns the full pathname of a file distributed with sampleModels

    .. doctest::

        >>> from cellnopt.data import cnodata
        >>> filename = cnodata("ToyModelPCB.sif") #doctest: +SKIP

    """
    if filename.endswith(".sif")==False and filename.endswith(".csv")==False:
        raise ValueError("webdata ease access to SIF or CSV format only (small caps extension)")

    if os.path.isfile(filename):
        return filename


    # by default, search for a file on the web but the use can overwrite that
    # behaviour
    if local == True:
        pass
    else:
        try:
            filename = _get_data(filename, verbose)
            # if it cannot be found on the web, try locally
        except IOError:
            try:
                filename = _webdata(filename, verbose)
            # if even locally it cannot be found, raise an error
            except Exception:
                raise Exception
    return filename


def _webdata(filename, verbose):
    """Search for a file in the samplemodels repository on the web.

    """
    # try all URL in baseurls
    for url in baseurls:
        url += os.sep + filename
        a = urllib.urlopen(url)

        if a.getcode() == 200:
            local_filename, obj = urllib.urlretrieve(url)
            logging.info("Found %s in %s" % (filename, url))
            return local_filename
        else:
            if verbose==True:
                import warnings
                logging.warning("Could not find %s in %s" % (filename, url))
            raise IOError


def _get_data(filename, verbose=True):
    if verbose:
        logging.basicConfig(level=logging.INFO)
    else:
        logging.basicConfig(level=logging.ERROR)

    try:
        logging.info("Trying local sampleModels")
        import sampleModels
        from sampleModels import get_data
        filename = get_data(filename, verbose)
        logging.info("Found with get_data from sampleModels")
        return filename
    except ImportError, e:
        print e
        logging.warning("sampleModels not installed")
    except:
        pass 

    try:
        logging.info("Trying to use cellnopt.wrapper.")
        from cellnopt.wrapper import get_data
        filename = get_data(filename)
        return filename
    except ImportError, e:
        print e
        logging.warning("cellnopt.wrapper not installed")
    except:
        raise IOError("file could not be found with sampleModels or cellnopt.wrapper.")


def checksum(filename):
    """Simple checksum on a cnodata file


    :param filename: a file to be found in sampleModels or cellnopt.wrapper

    >>> from cellnopt.data import checksum
    >>> cs = checksum(cnodata("PKN-ToyPCB.sif"))

    """
    import hashlib
    filename_path = cnodata(filename)
    checksum = hashlib.md5(file(filename_path, 'r').read()).digest()
    return checksum
