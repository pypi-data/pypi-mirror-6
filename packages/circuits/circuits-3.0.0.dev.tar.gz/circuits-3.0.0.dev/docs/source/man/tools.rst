Tools
=====


.. todo:: Write about this ...
