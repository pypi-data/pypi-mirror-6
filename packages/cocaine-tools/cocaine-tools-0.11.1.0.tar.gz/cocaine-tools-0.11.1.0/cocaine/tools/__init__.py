import logging

__author__ = 'EvgenySafronov <division494@gmail.com>'


log = logging.getLogger(__name__)