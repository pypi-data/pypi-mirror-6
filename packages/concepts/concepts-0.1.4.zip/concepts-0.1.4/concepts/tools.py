# tools.py

"""Tools."""

__all__ = ['lazyproperty']

class lazyproperty(object):
    """Non-data descriptor caching the computed result as instance attribute.

    >>> class Spam(object):
    ...     @lazyproperty
    ...     def eggs(self):
    ...         return 'spamspamspam'

    >>> spam=Spam(); spam.eggs
    'spamspamspam'

    >>> spam.eggs='eggseggseggs'; spam.eggs
    'eggseggseggs'

    >>> Spam().eggs
    'spamspamspam'
    """

    def __init__(self, fget):
        self.fget = fget
        for attr in ('__module__', '__name__', '__doc__'):
            setattr(self, attr, getattr(fget, attr))

    def __get__(self, instance, owner):
        if instance is None:
            return self
        result = instance.__dict__[self.__name__] = self.fget(instance)
        return result


def _test(verbose=False):
    import doctest
    doctest.testmod(verbose=verbose)

if __name__ == '__main__':
    _test()
