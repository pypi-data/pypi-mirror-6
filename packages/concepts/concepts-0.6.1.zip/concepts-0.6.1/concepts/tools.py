# tools.py

import csv
import codecs
from itertools import permutations, groupby, starmap
import operator

try:
    import cStringIO as StringIO
except ImportError:
    import StringIO

__all__ = ['maximal', 'lazyproperty', 'UnicodeReader', 'UnicodeWriter']


def maximal(iterable, comparison=operator.lt, _groupkey=operator.itemgetter(0)):
    """Yield the unique maximal elements from iterable using comparison.

    >>> list(maximal([1, 2, 3, 3]))
    [3]

    >>> list(maximal([1]))
    [1]
    """
    iterable = set(iterable)
    if len(iterable) < 2:
        return iterable
    return (item for item, pairs
        in groupby(permutations(iterable, 2), key=_groupkey)
        if not any(starmap(comparison, pairs)))


class lazyproperty(object):
    """Non-data descriptor caching the computed result as instance attribute.

    >>> class Spam(object):
    ...     @lazyproperty
    ...     def eggs(self):
    ...         return 'spamspamspam'

    >>> spam=Spam(); spam.eggs
    'spamspamspam'

    >>> spam.eggs='eggseggseggs'; spam.eggs
    'eggseggseggs'

    >>> Spam().eggs
    'spamspamspam'
    """

    def __init__(self, fget):
        self.fget = fget
        for attr in ('__module__', '__name__', '__doc__'):
            setattr(self, attr, getattr(fget, attr))

    def __get__(self, instance, owner):
        if instance is None:
            return self
        result = instance.__dict__[self.__name__] = self.fget(instance)
        return result


# from stdlib recipe

class UTF8Recoder(object):

    def __init__(self, f, encoding):
        self.reader = codecs.getreader(encoding)(f)

    def __iter__(self):
        return self

    def next(self):
        return self.reader.next().encode('utf8')


class UnicodeReader(object):

    def __init__(self, f, dialect=csv.excel, encoding='utf8', **kwargs):
        f = UTF8Recoder(f, encoding)
        self.reader = csv.reader(f, dialect=dialect, **kwargs)

    def next(self):
        row = self.reader.next()
        return [unicode(s, 'utf8') for s in row]

    def __iter__(self):
        return self


class UnicodeWriter(object):

    def __init__(self, f, dialect=csv.excel, encoding='utf-8', **kwargs):
        self.queue = StringIO.StringIO()
        self.writer = csv.writer(self.queue, dialect=dialect, **kwargs)
        self.stream = f
        self.encoder = codecs.getincrementalencoder(encoding)()

    def writerow(self, row):
        self.writer.writerow([s.encode('utf8') for s in row])
        # Fetch UTF-8 output from the queue ...
        data = self.queue.getvalue()
        data = data.decode('utf8')
        # ... and reencode it into the target encoding
        data = self.encoder.encode(data)
        # write to the target stream
        self.stream.write(data)
        # empty queue
        self.queue.truncate(0)

    def writerows(self, rows):
        for row in rows:
            self.writerow(row)
