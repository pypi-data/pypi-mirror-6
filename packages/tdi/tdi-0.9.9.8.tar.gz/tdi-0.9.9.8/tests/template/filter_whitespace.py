#!/usr/bin/env python
import warnings as _warnings
_warnings.resetwarnings()
_warnings.filterwarnings('error')

from tdi.tools import html as _html_tools
from tdi import html

html_compressed = html.replace(eventfilters=[_html_tools.MinifyFilter])

template = html_compressed.from_string("""
<html>
<head>
    <title>Boo! </title>
    <script>
        Some script
    </script>
    <style><!--
        Some style
    --></style>
</head>
<body>
    <p>Hello <b>YOU!</b> reader!

    Now...      <br />
    <form>
        <textarea   >Some
text
in     here.
        </textarea >
    </form>

    <!-- Comment! -->
    <pre>
        More text

            in


                here.
    </pre>
</body>
</html  >
abc

""")

template.render()
print
