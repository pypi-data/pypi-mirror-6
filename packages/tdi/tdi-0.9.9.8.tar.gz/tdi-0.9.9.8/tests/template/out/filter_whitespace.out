<html><head><title>Boo!</title><script>Some script</script><style><!--
        Some style
    --></style></head><body><p>Hello <b>YOU!</b> reader! Now... <br /><form><textarea>Some
text
in     here.
        </textarea></form><pre>
        More text

            in


                here.</pre></body></html>abc
